System/FilePath/Glob/Directory.hs

```diff
@@ -213,21 +214,57 @@ matchTypedAndGo opts (AnyDir n p:ps) path absPath = do
              p' = Pattern (unPattern p ++ [AnyNonPathSeparator])
 
          case unconditionalMatch || matchWith (matchOptions opts) p' path of
-              True | isDir  -> do
-                 contents <- getRecursiveContents
-                                (absPath ++ replicate n pathSeparator)
+              True | isDir -> do
+                 contents <- getRecursiveContents absPath
                  return $
                     -- foo**/ should match foo/ and nothing below it
                     -- relies on head contents == absPath
                     if null ps
-                       then (DL.singleton $ DL.head contents, DL.tail contents)
-                       else partitionDL (any m . pathParts) contents
-
-              True | m path -> return (DL.singleton absPath, DL.empty)
-              _             -> didn'tMatch opts path absPath isDir
+                       then ( DL.singleton $
+                                DL.head contents
+                                ++ replicate n pathSeparator
+                            , DL.tail contents
+                            )
+                       else let (matches, nonMatches) =
+                                   partitionDL fst
+                                      (fmap (recursiveMatch n m) contents)
+                             in (fmap snd matches, fmap snd nonMatches)
+
+              True | m path ->
+                 return ( DL.singleton $
+                             takeDirectory absPath
+                             ++ replicate n pathSeparator
+                             ++ path
+                        , DL.empty
+                        )
+              _ ->
+                 didn'tMatch opts path absPath isDir
 
 matchTypedAndGo _ _ _ _ = error "Glob.matchTypedAndGo :: internal error"
 
+recursiveMatch :: Int -> (FilePath -> Bool) -> FilePath -> (Bool, FilePath)
+recursiveMatch n isMatch path =
+   case find isMatch (pathParts path) of
+        Just matchedSuffix ->
+           case stripSuffix matchedSuffix path of
+                Just dir ->
+                   ( True
+                   , dir
+                     ++ replicate (n-1) pathSeparator
+                     ++ matchedSuffix
+                   )
+                Nothing ->
+                   -- the fact that we won't hit this branch relies on the
+                   -- property that every element of `pathParts fp` is a suffix
+                   -- of `fp`.
+                   error "Glob.recursiveMatch :: internal error"
```

Comment:
Since this can't happen, I would take advantage of it and optimize `stripSuffix` to something along the lines of `\suffix path -> take (length path - length suffix) path`.